### ISSUE TITLE
Refactor/reorganize UpgradeRules

### ISSUE THREAD

**Issue by reaperrr** [OpenRA/OpenRA#9190]
While I can see why they are necessary, for me they are often responsible for the majority of not-automatically-solvable rebase conflicts, and additionally a major pain to navigate, due how large the file has become.

It would be nice if we found a solution to at least reduce rebase conflicts, as these are extremely annoying.
##

**Comment by pchote** [comment]
I have discussed this several times in IRC. The plan for a long time has been to split upgrade rules into an interface / plumbing and individual rule classes, so that mods can define their own rules and completely avoid rebase conflicts.

**Comment by penev92** [comment]
OK, after some discussion on IRC I am posting the results here for future everyone's reference and in case someone disagrees with the plan:

1. Add `IUpgradeRules`.
2. Have `UpgradeRules` implement that (custom mods can have their own classes that implement it).
3. Create one `UpgradeRules<date>` class in Common implementing `IUpgradeRules` after each release (feature freeze) and use it for the next cycle.
(So far we've dealt with the huge file + we have support for custom mods having their own upgrade rules - useful for making maps compatible with the mod)
 
4. Add an `UpgradeRulesBeforeAttribute` with a Date property.
5. Create a new method for each upgrade rule and put an `[UpgrageRulesBefore(<date>)]` attribute on it. (*Don't forget to put a descriptive comment on the method about its purpose!*)
(This will (hopefully) eliminate the constant merge conflicts and need to rebase because of upgrade rules)
 
6. Have the utility get all classes implementing `IUpgradeRules`.
7. From those classes get all methods with `UpgradeRulesBefore date > engineVersion`.
8. Pass the ID of the mod that is being upgraded (selected by the user as a parameter for the utility) to each of those methods, as some of them may want to run only for specific mods (especially in the presence of custom mods with their own upgrade rules).

While a class per rule will ensure no more rebase conflicts, we'd be looking at tens of new files withing weeks, which doesn't sound like a good idea to me. A new class per release cycle will keep the files small enough to manage, and will mean around 2-3 new files per year at our current speed.

**Comment by GraionDilach** [comment]
I like the idea - honestly I always thought upgrade rules were mostly for non-bleed (playtest/release) modders, anyone picking up bleed to mod took the dangers of dealing with the catching up as part of that pick -, what sounds wrong alone to me is that it would be between releases. I'd prefer files per playtests, because this breaks OpenRA modmaps uploaded to the resource site during a release cycle pretty much as-is.

**Comment by penev92** [comment]
Nothing should change for maps on the resource site - its utility is always running the version of the last release, so it will be unaware of newer upgrade rules until it is automatically updated when the next release is tagged, just like now. (Someone correct me if I'm wrong please, it's quite late here now.)

**Comment by GraionDilach** [comment]
It runs on the last build - it can either be a playtest or a release. I've seen maps tagged with playtest compatibility for sure.

**Comment by LavenderMoon** [comment]
I think penev meant they'll be running the last Playtest or Proper release, not the latest git build.

**Comment by Phrohdoh** [comment]
My only issue with this is that custom mods may not be released with versions that are dated.
It may instead be `3.0-beta`. Of course the author just needs to note down the date of build/release, but it is still something to think about.

How is `UpgradeRulesBeforeAttribute` different from the other methods?
`Rules` is a generalization and everything falls under that category.

**Comment by penev92** [comment]
No, they are pretty different.
The 4 different attributes were suggested to reflect the current setup with `UpgradeActors()`, `UpgradeActorRules()`, `UpgradePlayers()` and `UpgradeWeaponRules()`. I didn't check the current names at the time so the names don't quite match, but that's the idea.

**Comment by jaimetotal** [comment]
I've started this refactoring on my own before checking this solution but I've reached an obstacle with the proposed design or perhaps I've misunderstood.

By having multiple methods, for each `Upgrade*Attribute()` in each class that implements `IUpgradeRules`, how can I guarantee the method signature when using `Invoke()` with Reflection?

Or the proposed offer is to have one method by class, with multiple attributes? This design has the issue that the date would be repeated multiple times. Instead, it's best that it's a property of `UpgradeRules`

The current state looks like this:
`
internal class Upgrade20150430 : IUpgradeRules
    {

        [UpgradeActorBefore(20150430)]
        public void Upgrade(Action<string> emitError, Action<string> emitWarning, MiniYamlNode node)
        {
            if (node.Key == "Health")
                UpgradeRulesUtils.ConvertFloatToIntPercentage(ref node.Value.Value);
        }

    }
`

**Comment by pchote** [comment]
The approach i've been arguing for would look something like this (using #8066 as an example):

```csharp
public abstract class UpgradeRule
{
	public abstract int Date { get; }
	public abstract string Name { get; }
	public abstract string Description { get; }
	
	public void ReportManualChange(string message)
	{
		// TODO: do something with this
	}
	
	public virtual void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode> nodes, MiniYamlNode parent, int depth) { /* do nothing */ }
	// TODO: and so on for the other types
}

public class ReplaceDeathTypes : UpgradeRule
{
	public override int Date { get { return 20150524; } }
	public override string Name { get { return "Replace generic death type definitions with the new DamageTypes code."; } }
	public override string Description
	{
		get
		{
			return "The warhead DeathType index has been removed, and death effects are now " + 
				"based on the DamageTypes list.  Remove DeathType from all warheads and add "
				"the DeathTypes mapping to WithDeathAnimation / DeathSounds.";
		}
	}
	
	static void RenameDamageTypes(MiniYamlNode damageTypes)
	{
		// TODO: We would probably want to define one rule per mod, instead of hardcoding them all here
		var mod = Game.ModData.Manifest.Mod.Id;
		if (mod == "cnc" || mod == "ra")
		{
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType1", "DefaultDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType2", "BulletDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType3", "SmallExplosionDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType4", "ExplosionDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType5", "FireDeath");
		}

		if (mod == "cnc")
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType6", "TiberiumDeath");

		if (mod == "ra")
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType6", "ElectricityDeath");

		if (mod == "d2k")
		{
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType1", "ExplosionDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType2", "SoundDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType3", "SmallExplosionDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType4", "BulletDeath");
		}

		if (mod == "ts")
		{
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType1", "BulletDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType2", "SmallExplosionDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType3", "ExplosionDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType5", "FireDeath");
			damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType6", "EnergyDeath");
		}
	}
	
	public override void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode> nodes, MiniYamlNode parent, int depth)
	{
		// Replace numbers with strings for DeathSounds.DeathType
		if (node.Key.StartsWith("DeathSounds"))
		{
			var deathTypes = node.Value.Nodes.FirstOrDefault(x => x.Key == "DeathTypes");
			if (deathTypes != null)
			{
				var types = FieldLoader.GetValue<string[]>("DeathTypes", deathTypes.Value.Value);
				deathTypes.Value.Value = string.Join(", ", types.Select(type => "DeathType" + type));

				RenameDamageTypes(deathTypes);
			}
		}
		
		ReportManualChange("Define WithDeathAnimation DeathTypes mapping.");
	}
	
	public override void UpgradeWeaponRules(int engineVersion, ref List<MiniYamlNode> nodes, MiniYamlNode parent, int depth)
	{
		// Replace "DeathTypeX" damage types with proper words
		if (node.Key.StartsWith("Warhead") && node.Value.Value == "SpreadDamage")
		{
			var damageTypes = node.Value.Nodes.FirstOrDefault(x => x.Key == "DamageTypes");
			if (damageTypes != null)
				RenameDamageTypes(damageTypes);
		}
	}
}
```

**Comment by GraionDilach** [comment]
Wait, @pchote, what happens with third-aprty mods? If you intend to add the ids to each upgrade rule then there should be atleast a fallback for them otherwise the upgrade rules become pointless completely.

**Comment by pchote** [comment]
Note the
>  // TODO: We would probably want to define one rule per mod, instead of hardcoding them all here

The rules above were copied from #8066, and for the purposes of illustration it wasn't worth the time to actually make that change.

**Comment by jaimetotal** [comment]
The only downside I see with that approach is that I have to expose 4 methods (if it's only maps) or 8 methods (all upgrades) that may not be always used.

This is our current upgrades rules:

- Actors
- Players
- ActorRules
- WeaponRules
- Tileset
- Cursors
- ChromeMetrics
- ChromeLayout

If I were to add a new kind of upgrade, let's say `ChromeCursors` (making this up), I would have to expose the methods to old upgrades that didn't required it. Unless this **never **happens.

**Comment by pchote** [comment]
The upgrade rule methods would be virtual but not abstract.  Only the rules that actually use them would need to override them.

**Comment by jaimetotal** [comment]
Yes, but when fetching them by reflection, I would run all of them, even if they weren't override.
By using interfaces, I don't need to call so many.

Like this:

```c#
public interface IUpgradeRule
{
    int Date { get; }
    string Name { get; }
    string Description { get; }
}

public interface IUpgradeActorRules : IUpgradeRule
{
	public virtual void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode> nodes);

	
public interface IUpgradeWeaponRules : IUpgradeRule
{
	public virtual void UpgradeWeaponRules(int engineVersion, ref List<MiniYamlNode> nodes, MiniYamlNode parent, int depth);
}

public class ReplaceDeathTypes : IUpgradeActorRules, IUpgradeWeaponRules
{
    public override int Date { get { return 20150524; } }
    public override string Name { get { return "Replace generic death type definitions with the new DamageTypes code."; } }
    public override string Description
    {
        get
        {
            return "The warhead DeathType index has been removed, and death effects are now " + 
                "based on the DamageTypes list.  Remove DeathType from all warheads and add "
                "the DeathTypes mapping to WithDeathAnimation / DeathSounds.";
        }
    }

    static void RenameDamageTypes(MiniYamlNode damageTypes)
    {
        // TODO: We would probably want to define one rule per mod, instead of hardcoding them all here
        var mod = Game.ModData.Manifest.Mod.Id;
        if (mod == "cnc" || mod == "ra")
        {
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType1", "DefaultDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType2", "BulletDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType3", "SmallExplosionDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType4", "ExplosionDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType5", "FireDeath");
        }

        if (mod == "cnc")
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType6", "TiberiumDeath");

        if (mod == "ra")
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType6", "ElectricityDeath");

        if (mod == "d2k")
        {
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType1", "ExplosionDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType2", "SoundDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType3", "SmallExplosionDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType4", "BulletDeath");
        }

        if (mod == "ts")
        {
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType1", "BulletDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType2", "SmallExplosionDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType3", "ExplosionDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType5", "FireDeath");
            damageTypes.Value.Value = damageTypes.Value.Value.Replace("DeathType6", "EnergyDeath");
        }
    }

    public void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode> nodes)
    {
        // Replace numbers with strings for DeathSounds.DeathType
        if (node.Key.StartsWith("DeathSounds"))
        {
            var deathTypes = node.Value.Nodes.FirstOrDefault(x => x.Key == "DeathTypes");
            if (deathTypes != null)
            {
                var types = FieldLoader.GetValue<string[]>("DeathTypes", deathTypes.Value.Value);
                deathTypes.Value.Value = string.Join(", ", types.Select(type => "DeathType" + type));

                RenameDamageTypes(deathTypes);
            }
        }

        ReportManualChange("Define WithDeathAnimation DeathTypes mapping.");
    }

    public void UpgradeWeaponRules(int engineVersion, ref List<MiniYamlNode> nodes, MiniYamlNode parent, int depth)
    {
        // Replace "DeathTypeX" damage types with proper words
        if (node.Key.StartsWith("Warhead") && node.Value.Value == "SpreadDamage")
        {
            var damageTypes = node.Value.Nodes.FirstOrDefault(x => x.Key == "DamageTypes");
            if (damageTypes != null)
                RenameDamageTypes(damageTypes);
        }
    }
}
```

**Comment by pchote** [comment]
Unfortunately there's a lot of context missing from this ticket because most of the recent discussion about this feature was done on IRC or not documented at all.

Interfaces are cleaner if you want to run each set of upgrade types in isolation (e.g. all actor rules, then all weapon rules, then all tileset rules, etc), which is what the current setup does.  For the new system we want to enumerate by upgrade rule so that the rule output is sensible (listing each rule that is being applied, and then any manual steps that it may require), and also be able to take a copy of the yaml between each upgrade step so we can do a partial upgrade and save sane state if one of the rules fails (this then lets the modder manually fix their rules, and then run the command again to complete the upgrade).  This removes the benefit of the interfaces because we're already enumerating over all the rule objects - calling a no-op upgrade method isn't a big deal since performance isn't a concern, and this then reduces boilerplate.

**Comment by jaimetotal** [comment]
Ok, later I will revise the design and update the suggestion.

**Comment by pchote** [comment]
Reverse-duplicate of #14600 and fixed by #14964.  I knew that I had written about this before!
